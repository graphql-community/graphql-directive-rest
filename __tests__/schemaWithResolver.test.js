const { graphql } = require('graphql');
const nock = require('nock');
const directive = require('../index');
const schema = require('../example/schemaWithResolvers');

beforeAll(() => {
  nock.disableNetConnect();
});

afterEach(() => {
  nock.cleanAll();
});

afterAll(() => {
  nock.enableNetConnect();
});

test('getDirectiveDeclaration should be defined', () => {
  expect(directive.getDirectiveDeclaration()).toMatchSnapshot();
});

test('path is return correct result and fillParametersFromArgs replace parameters in url', async () => {
  nock('https://yesno.wtf:443')
    .get('/api')
    .reply(200, { answer: 'yes' });

  nock('https://randomuser.me:443', { encodedQueryParams: true })
    .get('/api/')
    .query({ gender: 'female' })
    .reply(200, {
      results: [{ gender: 'female', email: 'diana.marshall@example.com' }],
    });

  const response = await graphql(
    schema,
    `
      query($gender: String) {
        me(gender: $gender) {
          gender
          email
          admin
        }
      }
    `,
    {},
    {},
    { gender: 'female' }
  );

  expect(response).toMatchSnapshot();
});

test('return plain response', async () => {
  nock('https://api.github.com:443', { encodedQueryParams: true })
    .get('/users')
    .reply(200, [
      { login: 'ejo', avatar_url: 'https://avatar.example' },
      { login: 'ajo', avatar_url: 'https://avatars0.example' },
      { login: 'bojo', avatar_url: 'https://avatars0.example' },
    ]);

  const response = await graphql(
    schema,
    `
      query {
        users {
          login
          avatar_url
        }
      }
    `
  );

  expect(response).toMatchSnapshot();
});

test('if replace url params', async () => {
  nock('https://api.github.com:443', { encodedQueryParams: true })
    .get('/users/ejo')
    .reply(200, {
      login: 'ejo',
      avatar_url: 'https://avatars0.example',
    });

  const response = await graphql(
    schema,
    `
      query($user: String) {
        user(user: $user) {
          login
          avatar_url
        }
      }
    `,
    {},
    {},
    { user: 'ejo' }
  );

  expect(response).toMatchSnapshot();
});
