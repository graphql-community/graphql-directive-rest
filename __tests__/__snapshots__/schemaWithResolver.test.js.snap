// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`getDirectiveDeclaration should be defined 1`] = `
GraphQLDirective {
  "args": Array [
    Object {
      "astNode": undefined,
      "defaultValue": undefined,
      "description": null,
      "name": "url",
      "type": "String",
    },
    Object {
      "astNode": undefined,
      "defaultValue": "",
      "description": null,
      "name": "extractFromResponse",
      "type": "String",
    },
  ],
  "astNode": undefined,
  "description": undefined,
  "locations": Array [
    "FIELD_DEFINITION",
  ],
  "name": "rest",
}
`;

exports[`if replace url params 1`] = `
Object {
  "data": Object {
    "user": Object {
      "avatar_url": "https://avatars0.example",
      "login": "ejo",
    },
  },
}
`;

exports[`if throw error 1`] = `
Object {
  "data": Object {
    "user": Object {
      "avatar_url": null,
      "login": null,
    },
  },
}
`;

exports[`path is return correct result and fillParametersFromArgs replace parameters in url 1`] = `
Object {
  "data": Object {
    "me": Object {
      "admin": "yes",
      "email": "diana.marshall@example.com",
      "gender": "female",
    },
  },
}
`;

exports[`return plain response 1`] = `
Object {
  "data": Object {
    "users": Array [
      Object {
        "avatar_url": "https://avatar.example",
        "login": "ejo",
      },
      Object {
        "avatar_url": "https://avatars0.example",
        "login": "ajo",
      },
      Object {
        "avatar_url": "https://avatars0.example",
        "login": "bojo",
      },
    ],
  },
}
`;
